Adding test for stride and write method for RadClass

stompsjo opened the pull request:
Includes a test to ensure the `stride` variable is working correctly in RadClass. The test compares the length of the results and compares it to the expected number of samples when using a specified stride and integration. Resolves #13.

`stride` should advance the working timestamp, which varies independently of `integration`. This allows the integration window to overlap or be spaced out.

`RadClass` is also tweaked slightly, removing the `store_data` member variable. Instead, a `write()` method is included. Therefore, writing to file will no longer be automatic, but can be controlled by the user.

Commit:
adding test for stride in RadClass

gonuke reviewed (changes requested):
You should be able to test this without writing the CSV file.  You already have the data available in the RadClass object.

On a related note, I just realized that `run_all()` is where you also write data to the CSV file.  That would be better as a separate method, perhaps meaning that you don't need a `save_data` member variable.  If you had a memory issue that required you to save data as you process it (I don't think that's the case here, but maybe you are anticipating it?), things would be a little different, but you still might want to have it in 2 methods.

Finally, there is no way for the user to specify a different name for `results.csv`

Commit:
adding write function

Adding a write test

stompsjo commented:
> You should be able to test this without writing the CSV file. You already have the data available in the RadClass object.

Correct, I have updated each test case accordingly.

> On a related note, I just realized that `run_all()` is where you also write data to the CSV file. That would be better as a separate method, perhaps meaning that you don't need a `save_data` member variable. If you had a memory issue that required you to save data as you process it (I don't think that's the case here, but maybe you are anticipating it?), things would be a little different, but you still might want to have it in 2 methods.

Good point. I think a while ago I had envisioned `RadClass` being used with multiple files. The user would provide a list of filenames that would all be processed by the same object and therefore their results would be saved automatically. In practice, each `RadClass` object tends to only handle one file, or at least one file at a time. Thus, I think the user probably would prefer control over what is and isn't written. This has been updated.

> Finally, there is no way for the user to specify a different name for `results.csv`

Also updated, the implemented `write()` method takes a `filename` as an input.

stompsjo commented:
Since this PR is small, I think it is okay to expand scope to include the introduction of the `write()` method.

gonuke commented on tests/test_RadClass.py:
```diff
@@ -91,14 +86,52 @@ def test_cache():
+    #   counts * integration / live-time
+    expected = spectra * integration / test_data.livetime
+    expected_samples = int(test_data.timesteps /
+                           (integration + (stride - integration)))
```
Isnt' this just `stride`?  `(integration + (stride - integration)) `

gonuke commented on tests/test_RadClass.py:
```diff
@@ -91,14 +86,52 @@ def test_cache():
+    expected = spectra * integration / test_data.livetime
+    expected_samples = int(test_data.timesteps /
+                           (integration + (stride - integration)))
+    np.testing.assert_almost_equal(classifier.storage.iloc[0],
```
Isn't it important to tests the 2nd result to make sure the stride worked?  The first result will be the same independent of the stride.

stompsjo replied:
I have changed it to test the second result. I should note that because the `spectra` matrix is completely identical (i.e. each `[ij]` is 10.0), the value will be the same for each result. The first `assert` is just checking to make sure the integration math is still correct when stride != integration (similar to `test_integration`) while the second `assert` ensures the expected number of results when stride != integration. Perhaps it would be more interesting to have a `spectra` matrix that is not identical, but I'm not sure how to construct tests for a random matrix...

gonuke replied:
I agree that a non-uniform spectra matrix is a good idea

stompsjo replied:
Since `spectra` is global here, would it be sufficient enough to manually sum the correct indices in the test (using something like `np.sum()`) and compare that to what happens in `RadClass` or would that be too similar to `RadClass`'s own operations? This would have to be changed for several of the tests in `test_RadClass`.

gonuke replied:
For a known spectra you could simply pre-calculate the correct answer to the sum.

e.g. `spectra[i] = i` then 
`sum(spectra,j,k) = [(k+1)k - j(j-1)]/2` (or something like that)

Commit:
correcting math and testing in test_stride

redoing spectra from uniform to sequential

updating RadClass tests with new spectra

changing expected integration value calculation

stompsjo commented:
I have changed the way that the spectra is constructed as you suggested. Now it is an `nxm` matrix with `A_nm = n`. `expected` is now calculated using the sum formula `n(n+1)/2` with a correction for indexing being `n(n-1)/2`. I agree that this PR should be merged before #16.

It also looks like this may require a rebase, I will do this once we are ready to merge.

gonuke commented:
This looks good to me now.  I think it needs a rebase before tests can run

Commit:
adding integration/average correction to expected test values

stompsjo commented:
Rebase complete and tests are passing. I should note that I had to add a correction to the `expected` values for #19. Since `collapse_rows` now does averaging (i.e. divides by `integration`), so too must the `expected` values.

gonuke commented:
LGTM
